zip-0244.rst

```diff
@@ -1,8 +1,382 @@
+field values of transparent inputs to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdPrevoutHash"
+
+1b: ``sequence_digest``
+'''''''''''''''''''''''
+A BLAKE2b-256 hash of the 32-bit little-endian representation of all ``nSequence``
+field values of transparent inputs to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSequencHash"
+
+1c: ``outputs_digest``
+''''''''''''''''''''''
+A BLAKE2b-256 hash of the field encoding of all transparent outputs 
+belonging to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdOutputsHash"
+
+2: ``sprout_digest``
+`````````````````````````
+A BLAKE2b-256 hash of the field encoding of all Sprout ``JoinSplit`` components of the
+transaction followed by the field encoding of the ``joinSplitPubKey`` value. This is
+equivalent to the internal hash that is specified by ZIP 143 [#zip-0143]_ over the Sprout
+data, with the distinction that the unmodified hash finalized immediately after
+initialization with the personalization string is used in the case that no ``JoinSplit``
+components exist, rather than the ``uint256`` of ``0x0000...0000``.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdJSplitsHash"
+
+3: ``sapling_digest``
+`````````````````````
+The digest of Sapling components is composed of two subtrees which are organized to 
+permit easy interoperability with the ``CompactBlock`` representation of Sapling data
+specified by the ZIP 307 Light Client Protocol [#zip-0307]_.
+
+This digest is a BLAKE2b-256 hash of the following values ::
+
+   * 3a. ``sapling_spends_digest`` (32-byte hash)
+   * 3b. ``sapling_outputs_digest`` (32-byte hash)
+   * 3c. ``valueBalance`` (64-bit signed little-endian)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSaplingHash"
+
+3a: ``sapling_spends_digest``
+''''''''''''''''''''''''''''''
+This digest is a BLAKE2b-256 hash of the following values ::
+
+   * 3a.i.   ``sapling_spends_compact_digest`` (32-byte hash)
+   * 3b.ii.  ``sapling_spends_noncompact_digest`` (32-byte hash)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSSpendsHash"
+
+3a.i: ``sapling_spends_compact_digest``
+.......................................
+A BLAKE2b-256 hash of the field encoding of all nullifier field
+values of Sapling shielded spends belonging to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSSpendCHash"
+
+3a.ii: ``sapling_spends_noncompact_digest``
+...........................................
+A BLAKE2b-256 hash of the non-nullifier information for all Sapling shielded spends
+belonging to the transaction. For each spend, the following elements are included
+in the hash:: 
+
+   * 3a.ii.1 ``cv``      (field encoding bytes)
+   * 3a.ii.2 ``anchor``  (field encoding bytes)
+   * 3a.ii.3 ``rk``      (field encoding bytes)
+   * 3a.ii.4 ``zkproof`` (field encoding bytes)
```

Comment:
Why does this include `zkproof` which is a witness element?

Reply:
Oh, I guess because the signature is over the proof? This depends on whether we count it as a malleability attack if you malleate the proof, then re-sign the transaction. Notice that there are private inputs to a spend proof that are not necessarily known by everyone who can sign (for example, `path` and `nsk`). Also need to consider the threshold multisig case using FROST.

Reply:
Does the proof count as effecting data? If not, it's straightforward to move this to the signature hash.

Reply:
We've debated this at various points, and gone back and forward on what should be where 😄

Hmm... thinking about it on the spot, I think I now believe that proofs should not count as effecting data. The proof checks the correctness of a series of public commitments to the private effects. As long as those commitments are part of the effects, and the commitments correctly commit to all effects, I think we should be fine to exclude the proofs.

Reply:
I went to implement this change, and realized that one thing that changes if we move zkproof to the signature hash is that we will not achieve the goal that we'd previously stated of being able to just sign the txid in the case that there are no transparent inputs. However, I presume that we want the signature hash to cover the proofs; is that earlier stated goal incorrect, or is there another alternative here that I'm missing?

Reply:
This is complicated, but strictly speaking the following is a potential malleability attack against Sapling. Consider an adversary who has all the information needed to re-sign, but does not have nsk. They strip the signatures, malleate the proof (which is possible for Groth16), and re-sign. Now the tx has a different txid.

You can argue that this isn't a particularly serious attack because the separation between nsk and other keys isn't really relied on or particularly useful (and we're considering removing nsk from Orchard for that reason). But I can't rule out that there might be other attacks based on malleating proofs.

Reply:
Added as https://github.com/zcash/zips/pull/433/commits/becda9c54391f7adaacd1cdf23ca965cbb771fd6

Reply:
@ebfull and I discussed this in our pairing again, and decided on two things:
- The proofs should not count as effecting data.
- We should not sign the proofs.

This gives us the property that signatures of purely-shielded transactions are just signatures over the txid, while also removing any kind of txid malleability concern due to proofs.

@daira:
> This is complicated, but strictly speaking the following is a potential malleability attack against Sapling. Consider an adversary who has all the information needed to re-sign, but does not have nsk. They strip the signatures, malleate the proof (which is possible for Groth16), and re-sign. Now the tx has a different txid.

This doesn't occur if proofs are not part of the txid.